### PR TITLE
Generic matchers

### PR DESCRIPTION
**Pull request by apelisse**

### CODE REVIEW

**Comment by apelisse** [comment]
> +839 −1,310

This is removing almost 500 lines of code :-)

**Comment by grodrigues3** [comment]
Can we split the big commit into two smaller ones.  Namely,
1. One that creates `matcher.go, item.go, items.go`
1. One that updates all the dependencies.

**Comment by crimsonfaith91** [comment]
Great job! The generic matcher handles all 3 cases: `IssueComment`, `PullRequestComment`, and `IssueEvent`. :)

**Comment by crimsonfaith91** [comment]
Addressing #2806

**Comment by apelisse** [comment]
> Can we split the big commit into two smaller ones. Namely,
> 
> One that creates matcher.go, item.go, items.go
> One that updates all the dependencies.

Do you mean for review? Because splitting in two like this would break the build

**Comment by crimsonfaith91** [comment]
I think breaking it into 2 reviews will work great! :)

**Comment by apelisse** [comment]
Took me a while to understand but OK here's what I'll do:
- Add the new matcher code in one pull-request so that you can start using it to do something else
- Create a second PR to remove the old comment/event matcher code and update the code that uses it. We might need to be more careful when we load that pr.

Thanks,

**Comment by apelisse** [comment]
OK I've updated the code, we can merge that without having to change anything else.

File: mungegithub/mungers/matchers/items.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	return i
+}
+
+// AddEvents returns a new list with the added events
+func (i Items) AddEvents(events ...*github.IssueEvent) Items {
+	for _, event := range events {
+		i = append(i, (*Event)(event))
+	}
+
+	sort.Sort(i)
```
Comment (grodrigues3) [inline mungegithub/mungers/matchers/items.go]: verify sort order

File: mungegithub/mungers/matchers/matchers.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+	return false
+}
+
+func (a addLabelMatcher) MatchReviewComment(review *github.PullRequestComment) bool {
+	return false
+}
+
+// LabelName searches for event whose label starts with the string
+type LabelName string
+
```
Comment (grodrigues3) [inline mungegithub/mungers/matchers/matchers.go]: add compile check

File: mungegithub/mungers/matchers/matchers.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+func (l LabelName) MatchReviewComment(review *github.PullRequestComment) bool {
+	return false
+}
+
+// LabelPrefix searches for event whose label starts with the string
+type LabelPrefix string
+
+// Match if the label starts with the string
```
Comment (grodrigues3) [inline mungegithub/mungers/matchers/matchers.go]: again compile check

File: mungegithub/mungers/matchers/notification_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		if !reflect.DeepEqual(actualNotif, test.notif) {
+			t.Error(actualNotif, "doesn't match expected notif:", test.notif)
+		}
+	}
+}
+
+func TestStringNotification(t *testing.T) {
+	tests := []struct {
+		notif *Notification
+		str   string
```
Comment (grodrigues3) [inline mungegithub/mungers/matchers/notification_test.go]: should be called expected

File: mungegithub/mungers/matchers/items.go (hunk 1 above)
Comment (apelisse) [inline mungegithub/mungers/matchers/items.go]: Done.

File: mungegithub/mungers/matchers/matchers.go (hunk 2 above)
Comment (apelisse) [inline mungegithub/mungers/matchers/matchers.go]: Done

File: mungegithub/mungers/matchers/matchers.go (hunk 3 above)
Comment (apelisse) [inline mungegithub/mungers/matchers/matchers.go]: Done

**Comment by grodrigues3** [comment]
@crimsonfaith91 framework is updated and merged.  please use it for the notify and reassign

**Comment by crimsonfaith91** [comment]
@grodrigues3 Thank you very much! Let see whether i can get the inactive munger completed today. :)

**Pull request merged**